en/news/_posts/2019-12-06-separation-of-positional-and-keyword-arguments-in-ruby-3-0.md

```diff
@@ -0,0 +1,341 @@
+# Ruby 2.7: warning: The last argument is split into positional and keyword parameters
+#           warning: for `bar' defined here
+# Ruby 3.0: ArgumentError
+```
+
+### 2. Double splat with an empty hash (`**{}`)  passes no arguments
+
+In Ruby 2.6 or before, passing `**empty_hash` passes an empty Hash as a positional argument.  In Ruby 2.7 or later, it passes no arguments.
+
+```
+def foo(*args)
+  args
+end
+
+empty_hash = {}
+foo(**empty_hash)
+  #=> Ruby 2.6 or before: [{}]
+  #=> Ruby 2.7 or later: []
+```
+
+Note that `foo(**{})` passes nothing in both Ruby 2.6 and 2.7.  In Ruby 2.6 and before, `**{}` is removed by the parser, and in Ruby 2.7 and above, it is treated the same as `**empty_hash`, allowing for an easy way to pass no keyword arguments to a method.
+
+In Ruby 2.7, when calling a method with an insufficient number of required positional arguments, `foo(**empty_hash)` passes an empty hash with a warning emitted, for compatibility with Ruby 2.6.  This behavior will be removed in 3.0.
+
+```
+def foo(x)
+  x
+end
+
+empty_hash = {}
+foo(**empty_hash)
+  #=> Ruby 2.6 or before: {}
+  #=> Ruby 2.7: warning: The keyword argument is passed as the last hash parameter
+  #             warning: for `foo' defined here
+  #=> Ruby 3.0: ArgumentError: wrong number of arguments
+```
+
+### 3. The no-keyword-arguments syntax (`**nil`) is introduced
+
+You can use `**nil` in a method definition to explicitly mark the method accepts no keyword arguments. Calling such methods with keyword arguments will result in an `ArgumentError`. (This is actually a new feature, not an incompatibility)
+
+```
+def foo(*args, **nil)
+end
+
+foo(k: 1)
+  #=> Ruby 2.7 or later: no keywords accepted (ArgumentError)
+```
+
+This is useful to make it explicit that the method does not accept keyword arguments.  Otherwise, the keywords are absorbed in the rest argument in the above example.  If you extend a method to accept keyword arguments, the method may have incompatibility as follows:
+
+```
+# If a method accepts rest argument and no `**nil`
+def foo(*args)
+  p args
+end
+
+# Passing keywords are converted to a Hash object (even in Ruby 3.0)
+foo(k: 1) #=> [{:k=>1}]
+
+# If the method is extended to accept a keyword
+def foo(*args, mode: false)
+  p args
+end
+
+# The existing call may break
+foo(k: 1) #=> ArgumentError: unknown keyword k
+```
+
+## Why we're deprecating the automatic conversion
+
+The automatic conversion initially appeared to be a good idea, and worked well in many cases.  However, it had too many corner cases, and we have received many bug reports about the behavior.
+
+Automatic conversion does not work well when a method accepts optional positional arguments and keyword arguments.  Some people expect the last Hash object to be treated as a positional argument, and others expect it to be converted to keyword arguments.
+
+Here is one of the hardest cases:
+
+```
+def foo(x=1, **kwargs)
+  p [x, kwargs]
+end
+
+foo({}, **{}) #=> ?
+```
```

Comment:
This example is unconvincing for me because it uses `**{}` which almost no real Ruby code uses today. I think a simpler example using e.g. an optional argument and showing both possible desired outcomes would be much better.
I.e., if we did all those changes to save this I'd say not worth it.

Reply:
Actually, it is not just an artificial example.  I've changed the explanation to a more descriptive one in 874673ad.